lib/galaxy/webapps/galaxy/api/__init__.py

```diff
@@ -0,0 +1,84 @@
+"""
+This module *does not* contain API routes. It exclusively contains dependencies to be used in FastAPI routes
+"""
+from typing import (
+    Optional,
+)
+
+from fastapi import (
+    Cookie,
+    Depends,
+    Header,
+    HTTPException,
+    Query,
+)
+from sqlalchemy.orm import Session
+
+from galaxy import (
+    app as galaxy_app,
+    exceptions,
+    model,
+)
+from galaxy.app import UniverseApplication
+from galaxy.managers.session import GalaxySessionManager
+from galaxy.managers.users import UserManager
+from galaxy.model import User
+from galaxy.work.context import SessionRequestContext
+
+
+def get_app() -> UniverseApplication:
+    return galaxy_app.app
+
+
+def get_db(app: UniverseApplication = Depends(get_app)) -> Session:
+    # TODO: return sqlachemy 2.0 style session without autocommit and expire_on_commit!
+    return app.model.session
+
+
+def get_user_manager(app: UniverseApplication = Depends(get_app)):
+    return UserManager(app)
+
+
+def get_session_manager(app=Depends(get_app)) -> GalaxySessionManager:
+    # TODO: find out how to adapt dependency for Galaxy/Report/TS
+    return GalaxySessionManager(app.model)
+
+
+def get_session(session_manager: GalaxySessionManager = Depends(get_session_manager),
+                app: UniverseApplication = Depends(get_app),
+                galaxysession: Optional[str] = Cookie(None)) -> Optional[model.GalaxySession]:
+    if galaxysession:
+        session_key = app.security.decode_guid(galaxysession)
+        if session_key:
+            return session_manager.get_session_from_session_key(session_key)
+        # TODO: What should we do if there is no session? Since this is the API, maybe nothing is the right choice?
+
+
+def get_api_user(user_manager: UserManager = Depends(get_user_manager), key: Optional[str] = Query(None), x_api_key: Optional[str] = Header(None)) -> Optional[User]:
+    api_key = key or x_api_key
+    if not api_key:
+        return None
+    try:
+        return user_manager.by_api_key(api_key=api_key)
+    except exceptions.AuthenticationFailed as e:
+        raise HTTPException(status_code=e.status_code, detail=str(e))
+
+
+def get_user(galaxy_session: Optional[model.GalaxySession] = Depends(get_session), api_user: Optional[User] = Depends(get_api_user)) -> Optional[User]:
+    if galaxy_session:
+        return galaxy_session.user
+    return api_user
+
+
+def get_trans(app=Depends(get_app), user: Optional[User] = Depends(get_user),
+              galaxy_session: Optional[model.GalaxySession] = Depends(get_session),
+              ) -> SessionRequestContext:
+    app.model.session.expunge_all()
+    return SessionRequestContext(app=app, user=user, galaxy_session=galaxy_session)
+
+
+def get_admin_user(trans: SessionRequestContext = Depends(get_trans), user_manager: UserManager = Depends(get_user_manager)):
```

Comment:
I'm not 100% sure if the return type can be `User` or if there is some specific `AdminUser` class or something like that.
```suggestion
def get_admin_user(trans: SessionRequestContext = Depends(get_trans), user_manager: UserManager = Depends(get_user_manager)) -> Optional[User]:
```

Reply:
There is a `BooststrapAdminUser` now, so this could be a typing.Union[User, BootstrapAdminUser].
I'm not 100% convinced it makes sense to annotate everything, since in this case I guess we'd have to import the `BooststrapAdminUser` just to create the type annotation. Which I think is fine in this case, but I could see this becoming a burden in other cases.

Reply:
This is really a good point. I haven't thought of this burden because I have used them only in a small project.
Should we come up with some kind of Type Annotation Guidelines for Galaxy? or are there any rules or best practices available for this matter?

Reply:
Yes, that is a good idea, but maybe we can freestyle it a bit while we're still getting to know the pros and cons?
My rough thinking is that for everything that we consider public (like API routes) we should annotate all parameters and return types. For things that are really dynamic we might make it optional ?

Reply:
That makes perfect sense. Totally agree!

Reply:
I think we should probably annotate an alternative admin transaction so we don't need to inject an admin user that isn't used? Or just reuse the old decorator to be honest.

One thing is we're annotating explicit concrete classes for everything and it is causing me some distress - we're creating very explicit dependencies and deep transitive dependencies. If we're going to use types internally for things like app and trans we should be using more precise interfaces that they implement and that can be imported without bringing all of say app's dependencies.